src/core/amount.js

```diff
@@ -113,32 +110,45 @@ Amount.is_valid_full = function(j) {
 };
 
 Amount.NaN = function() {
-  var result = new Amount();
-  result._value = new BigNumber(NaN); // should have no effect
+  const result = new Amount();
+  result._value = new Value(NaN)._value; // should have no effect
   return result;                      // but let's be careful
 };
 
 // be sure that _is_native is set properly BEFORE calling _set_value
 Amount.prototype._set_value = function(value, roundingMode) {
-  assert(value instanceof BigNumber);
-  this._value = value.isZero() && value.isNegative() ? value.negated() : value;
+  assert(value instanceof Object);
+
+  this._value = this._value = value.isZero() && value.isNegative() ?
+      value.negated()._value : value._value;
+  if (!this._value) {
+    this._value = value.isZero() && value.isNegative() ?
+      value.negated() : value;
+  }
   this.canonicalize(roundingMode);
   this._check_limits();
+
 };
 
 // Returns a new value which is the absolute value of this.
 Amount.prototype.abs = function() {
-  return this.clone(this.is_negative());
+
+  const val = (new IOUValue(this._value)).abs();
+  return this._copy(val);
+
 };
 
 Amount.prototype.add = function(addend) {
-  var addendAmount = Amount.from_json(addend);
+  const addendAmount = Amount.from_json(addend);
 
   if (!this.is_comparable(addendAmount)) {
     return new Amount(NaN);
   }
 
-  return this._copy(this._value.plus(addendAmount._value));
+  const thisValue = new IOUValue(this._value);
+  const addendValue = new IOUValue(addendAmount._value);
```

Comment:
You can just make `this._value` be stored as an instance of `XRPValue` or `IOUValue` so you don't have to do this every time.